Improve Asset Manager Reload behavior.

mbalfour-amzn opened the pull request:
## What does this PR do?

This PR is attempting to improve how Reload works with Asset Manager to make it a little more functional and intuitive.

* Changes AM::ReloadAsset() to return the newly-created asset, so that callers can use it to monitor the state of version getting reloaded.
* Changes AssetCommon::Reload() to immediately change the current asset to the one getting reloaded.
* Changes AssetInternal::GetAssetData() to get any queued reloaded data as well as currently-loaded data, though at a higher priority.

The net effect of these changes is that calling myAsset->Reload() will now change the myAsset reference into a queued state and clear its current set of data, as opposed to leaving it in a Ready state with valid data. The previous logic meant that myAsset wouldn't know anything about the reloading, and the caller would need to wait for OnAssetReloaded to *then* change myAsset to the new reloaded asset reference. 

The new logic is also consistent with what myAsset->QueueLoad() does, which changes myAsset to refer to the queued data, instead of leaving it in a NotLoaded state and requiring users to wait for OnAssetReady.

The more subtle change is the one to GetAssetData(). This changes how BlockUntilLoadComplete() works with reloaded assets. Previously, if you called myAsset->Reload(); myAsset->BlockUntilLoadComplete(); the second call wouldn't actually block because it would see that myAsset is currently loaded (since it is) and would return immediately. With this change, the call will get the reload asset data instead and will block until the reload is complete. 

The likely functional change that the GetAssetData() change would introduce is that any code attempting to call GetAsset() or BusConnect(assetId) for an asset that's currently reloading would no longer get the previously-loaded asset, and would instead get a reference to the asset that's in process of reloading. I'm honestly not sure which is the more desirable result, a case could be made for either behavior.

## How was this PR tested?

Right now, it hasn't been thoroughly tested, it's only been tested specifically in the context of reloading StreamingImage data in conjunction with some other code changes. I wanted to put this up for a draft review and discussion before looking further into how to test it.

mbalfour-amzn committed:
Improve Asset Manager Reload behavior.
* Changes AM::ReloadAsset() to return the newly-created asset, so that callers can use it to monitor the state of version getting reloaded.
* Changes AssetCommon::Reload() to immediately change the current asset to the one getting reloaded.
* Changes AssetInternal::GetAssetData() to get any queued reloaded data as well as currently-loaded data, though at a higher priority.

The net effect of these changes is that calling myAsset->Reload() will now change the myAsset reference into a queued state and clear its current set of data, as opposed to leaving it in a Ready state with valid data. The previous logic meant that myAsset wouldn't know anything about the reloading, and the caller would need to listen for OnAssetReloaded to *then* change myAsset to the new reloaded asset reference.

The more subtle change is the one to GetAssetData(). This changes how BlockUntilLoadComplete() works with reloaded assets. Previously, if you called myAsset->Reload(); myAsset->BlockUntilLoadComplete(); the second call wouldn't actually block because it would see that myAsset is currently loaded (since it is) and would return immediately. With this change, the call will get the reload asset data instead and will block until the reload is complete.

Signed-off-by: Mike Balfour <82224783+mbalfour-amzn@users.noreply.github.com>

amzn-mike commented:
I'm currently working on a hot-reloading issue and its leading me into changes somewhat close to this.  I want to spend a bit more time today to see where I end up so I can have a better sense of how these two changes will work together and if we need to make any adjustments and then Ill take another look at this.

lawsonamzn commented:
I'd be okay with this kind of change too.  The original implementation didn't really get baked in the fires of real world usage, just beware of side effects here and be aware that you'll still get the OnReloaded.   In addition, even though the m_Assets hash switches its ref immedately (watch for deadlocks!) existing other systems may have a ref to the old one still, until reload finishes

lawsonamzn commented:
Just to make it clear what I mean, suppose you have 2 different class instances, class1, and class2, both with a member variable of type Asset<FOO> m_myAsset;
**Old way**
```
class1 calls m_myAsset = assetmanager->GetAsset<Foo>(123);
... assetmanager does not find it in m_assets.  It creates an entry for it and starts an initial load job, refcount is 1 now
class2 calls m_myAsset = assetmanager->GetAsset<Foo>(123);
... assetmanager finds it in m_assets and returns it.    Refcount is 2.
... asset load job finishes.  Class1 and class2 get OnAssetReady().
...
... some time passes
...
class1 calls m_myAsset->reload();
assetmanager creates a reload job for it and puts it on the queue.  It creates a new asset data, but doesn't link it into m_assets yet.
...
assetmanager finishes reloading it and swaps m_assets to point at the loaded one
Assetmanager sends OnAssetReloaded to all listeners
The following happnens all on the same frame:
class1 gets OnAssetReloaded and does m_myAsset = newAsset;  refcount of old asset  drops to 1 and new rises to 1
class2 gets OnassetReloaded and does m_myAsset = newAsset;  refecount of old asset drops to 0 and dies, new rises to 2
```

Contrast with this "new way"
```
class1 calls m_myAsset = assetmanager->GetAsset<Foo>(123);
... assetmanager does not find it in m_assets.  It creates an entry for it and starts an initial load job, refcount is 1 now
class2 calls m_myAsset = assetmanager->GetAsset<Foo>(123);
... assetmanager finds it in m_assets and returns it.    Refcount is 2.
... asset load job finishes.  Class1 and class2 get OnAssetReady().
...
... some time passes
...
class1 calls m_myAsset = m_myAsset->reload();
refcount of the old data is immediately decremented by 1.
am immediately replaces it in m_assets with the new one.  new one is immediately incremented by 1.
job begins to reload it
...
there is now a gap where class1 has the new, not yet loaded data ptr in its m_myAsset, but class2 still has the old data.  This means there is now a disagreement between them, and this may or may not cuase problems depending on the system involved.
...
assetmanager finishes reloading it and swaps m_assets to point at the loaded one
Assetmanager sends OnAssetReloaded to all listeners
The following happnens all on the same frame:
class1 gets OnAssetReloaded and does m_myAsset = newAsset;  But they're the same ptr so nothing happens
class2 gets OnassetReloaded and does m_myAsset = newAsset;  refecount of old asset drops to 0 and dies, new rises to 2
```
At the end, both classes converge (Assuming they are listening to OnAssetReloaded correctly) but they do disagree in the interim.

The other side effects this change has
* If only one class has a refcount, Reload will immediately free memory since the refcount of the old will immediately drop to 0.  Make sure any fixups occur at that point not onassetreloaded!
* What happens to someone calling GetAsset<> during the gap before it finishes reloading?  Will they get an OnAssetReady? or will they never get it because they will instead get an OnAssetReloaded?

jackalbe commented on Code/Framework/AzCore/AzCore/Asset/AssetCommon.h:
```diff
@@ -1203,7 +1203,14 @@ namespace AZ
+                auto asset = AssetInternal::ReloadAsset(m_assetData, m_loadBehavior);
+
+                // Only assign on success so we don't lose the assetId/assetType if the above failed
+                if (asset)
```
if this fails, is it still valid to have the *this pointing to a failed reloaded asset?

mbalfour-amzn committed:
Merge branch 'development' into mbalfour/asset_manager_reload_fixes
Signed-off-by: Mike Balfour <82224783+mbalfour-amzn@users.noreply.github.com>

mbalfour-amzn committed:
Merge branch 'development' into mbalfour/asset_manager_reload_fixes
Signed-off-by: Mike Balfour <82224783+mbalfour-amzn@users.noreply.github.com>

mbalfour-amzn committed:
Change Reload to take in filter func.
This enables asserts to let us know if dependencies are loading incorrectly on Reload calls.

Signed-off-by: Mike Balfour <82224783+mbalfour-amzn@users.noreply.github.com>

mbalfour-amzn committed:
Improved error messaging in the cases where the asset hint contains a non-empty value.

Signed-off-by: Mike Balfour <82224783+mbalfour-amzn@users.noreply.github.com>

mbalfour-amzn committed:
Merge branch 'development' into mbalfour/asset_manager_reload_fixes
Signed-off-by: Mike Balfour <82224783+mbalfour-amzn@users.noreply.github.com>
